### PR TITLE
[RLlib] Issue #14022: Trajectory View API fails in MA-env where a new agent terminates right away (done=True right after initial obs).

### LINKED ISSUE #14022: [rllib] Trajectory view API broken for variable-agent envs
@sven1977 Copying from https://discuss.ray.io/t/potential-bug-in-trajectory-view-api-for-multiagent-envs/754/5. This is also a probably source for #13802

Training fails with internal errors when using the trajectory view API with variable-agent envs -- e.g. those where agents are added and removed dynamically during individual episodes. Confirmed on Ray 1.1.0 and 1.2.0.dev0, python 3.8.3, Ubuntu 20.04 with Pytorch/Tensorflow.

```
import gym
import random
import unittest

import ray

from ray.tune.registry import register_env
from ray.rllib.agents.pg import PGTrainer
from ray.rllib.examples.env.multi_agent import BasicMultiAgent

from ray.rllib.env.multi_agent_env import MultiAgentEnv
from ray.rllib.examples.env.mock_env import MockEnv

class BasicMultiAgent(MultiAgentEnv):
    """Env of N independent agents, each of which exits after 25 steps."""

    def __init__(self, num):
        self.agents = {}
        self.agentID = 0
        self.dones = set()
        self.observation_space = gym.spaces.Discrete(2)
        self.action_space = gym.spaces.Discrete(2)
        self.resetted = False
        
    def spawn(self):
        agentID = self.agentID
        self.agents[agentID] = MockEnv(25)
        self.agentID += 1
        return agentID

    def reset(self):
        self.agents = {}
        self.spawn()
        self.resetted = True
        self.dones = set()
        
        obs = {}
        for i, a in self.agents.items():
           obs[i] = a.reset()

        return obs

    def step(self, action_dict):
        obs, rew, done, info = {}, {}, {}, {}
        for i, action in action_dict.items():
            obs[i], rew[i], done[i], info[i] = self.agents[i].step(action)
            if done[i]:
                self.dones.add(i)

        if random.random() > 0.75:
           i = self.spawn()
           obs[i], rew[i], done[i], info[i] = self.agents[i].step(action)
           if done[i]:
              self.dones.add(i)

        if len(self.agents) > 1 and random.random() > 0.25:
           keys = list(self.agents.keys())
           key  = random.choice(keys)
           done[key] = True
           del self.agents[key]

        done["__all__"] = len(self.dones) == len(self.agents)
        return obs, rew, done, info

class TestMultiAgentEnv(unittest.TestCase):
    @classmethod
    def setUpClass(cls) -> None:
        ray.init(num_cpus=4)

    @classmethod
    def tearDownClass(cls) -> None:
        ray.shutdown()

    def test_train_multi_agent_cartpole_single_policy(self):
        n = 10
        register_env("basic_multi_agent",
                     lambda _: BasicMultiAgent({'num_agents': 10}))
        pg = PGTrainer(
            env="basic_multi_agent",
            config={
                "num_workers": 0,
                "framework": "torch",
            })
        for i in range(50):
            result = pg.train()
            print("Iteration {}, reward {}, timesteps {}".format(
                i, result["episode_reward_mean"], result["timesteps_total"]))
            if result["episode_reward_mean"] >= 50 * n:
                return
        raise Exception("failed to improve reward")

if __name__ == '__main__':
   TestMultiAgentEnv().test_train_multi_agent_cartpole_single_policy()
```

### PR DESCRIPTION
**Pull request by sven1977**
Issue #14022: Trajectory View API fails in MA-env where a new agent terminates right away (done=True right after initial obs).

Issue #14022
## Related issue number

Closes #14022
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jsuarez5341** [comment]
Not sure how your release structure works -- any chance this will make it into 1.2 for next week?

File: rllib/examples/env/multi_agent.py [hunk 1]
```diff
@@ -87,6 +88,62 @@ def step(self, action_dict):
...
 
 
+class FlexAgentsMultiAgent(MultiAgentEnv):
+    """Env of independent agents, each of which exits after n steps."""
+
+    def __init__(self):
+        self.agents = {}
+        self.agentID = 0
+        self.dones = set()
+        self.observation_space = gym.spaces.Discrete(2)
```
Comment (michaelzhiluo) [inline rllib/examples/env/multi_agent.py:98]: This shouldn't be hardcoded right?

File: rllib/examples/env/multi_agent.py [hunk 2]
```diff
@@ -87,6 +88,62 @@ def step(self, action_dict):
...
+    def step(self, action_dict):
+        obs, rew, done, info = {}, {}, {}, {}
+        # Apply the actions.
+        for i, action in action_dict.items():
+            obs[i], rew[i], done[i], info[i] = self.agents[i].step(action)
+            if done[i]:
+                self.dones.add(i)
+
+        # Sometimes, add a new agent to the episode.
+        if random.random() > 0.75:
```
Comment (michaelzhiluo) [inline rllib/examples/env/multi_agent.py:130]: What is the motivation behind killing/creating new agents during the episode?

**Review by michaelzhiluo — CHANGES_REQUESTED** [review]

File: rllib/examples/env/multi_agent.py (hunk 2 above)
Comment (jsuarez5341) [reply rllib/examples/env/multi_agent.py:130]: Applicability to an entire new class of environments, including artificial life inspired work. I'm the author of one such project in which this bug broke the trajectory API https://jsuarez5341.github.io/neural-mmo/build/html/rst/userguide.html

File: rllib/examples/env/multi_agent.py (hunk 1 above)
Comment (jsuarez5341) [reply rllib/examples/env/multi_agent.py:98]: This code was copied from my repro -- idea was just to have a setting where agents are periodically added/deleted. You could make the obs space not hard coded. As for the agentID, that's a counter that is incremented whenever a new agent is spawned such that each agent always has a unique key

File: rllib/examples/env/multi_agent.py (hunk 2 above)
Comment (sven1977) [reply rllib/examples/env/multi_agent.py:130]: @michaelzhiluo This is just an example env to simulate this kind of behavior in real envs.

Comment (michaelzhiluo) [reply rllib/examples/env/multi_agent.py:130]: Very cool! Aight approved.

**Review by michaelzhiluo — APPROVED** [review]

**Comment by sven1977** [comment]
No, this will not make it into 1.2 (already out at the time of this writing :/ ). Sorry!

**Pull request merged**